lib/class-wp-block-supports.php

```diff
@@ -0,0 +1,188 @@
+		$this->block_supports[ $block_support_name ] = array_merge(
+			$block_support_config,
+			array( 'name' => $block_support_name )
+		);
+	}
+
+
+	/**
+	 * Generates an array of HTML attributes, such as classes, by applying to
+	 * the given block all of the features that the block supports.
+	 *
+	 * @param  array $parsed_block Block as parsed from content.
+	 * @return array               Array of HTML attributes.
+	 */
+	public function apply_block_supports( $parsed_block ) {
+		$block_attributes = $parsed_block['attrs'];
+		$block_type       = WP_Block_Type_Registry::get_instance()->get_registered(
+			$parsed_block['blockName']
+		);
+
+		// If no render_callback, assume styles have been previously handled.
+		if ( ! $block_type || empty( $block_type ) ) {
+			return array();
+		}
+
+		$output = array();
+		foreach ( $this->block_supports as $name => $block_support_config ) {
+			if ( ! isset( $block_support_config['apply'] ) ) {
+				continue;
+			}
+
+			$new_attributes = call_user_func(
+				$block_support_config['apply'],
+				$block_type,
+				$block_attributes
+			);
+
+			if ( ! empty( $new_attributes ) ) {
+				foreach ( $new_attributes as $attribute_name => $attribute_value ) {
+					if ( empty( $output[ $attribute_name ] ) ) {
+						$output[ $attribute_name ] = $attribute_value;
+					} else {
+						$output[ $attribute_name ] .= " $attribute_value";
+					}
+				}
+			}
+		}
+
+		return $output;
+	}
+
+	/**
+	 * Registers the block attributes required by the different block supports.
+	 */
+	private function register_attributes() {
+		$block_registry         = WP_Block_Type_Registry::get_instance();
+		$registered_block_types = $block_registry->get_all_registered();
+		foreach ( $registered_block_types as $block_type ) {
+			if ( ! property_exists( $block_type, 'supports' ) ) {
+				continue;
+			}
+			if ( ! $block_type->attributes ) {
+				$block_type->attributes = array();
+			}
+
+			foreach ( $this->block_supports as $name => $block_support_config ) {
+				if ( ! isset( $block_support_config['register_attribute'] ) ) {
+					continue;
+				}
+
+				call_user_func(
+					$block_support_config['register_attribute'],
+					$block_type
+				);
+			}
+		}
+	}
+}
+
+/**
+ * Generates a string of attributes by applying to the current block being
+ * rendered all of the features that the block supports.
+ *
+ * @param array $extra_attributes Optional. Extra attributes to render on the block wrapper.
+ *
+ * @return string String of HTML classes.
+ */
+function get_block_wrapper_attributes( $extra_attributes = array() ) {
+	global $current_parsed_block;
+	$new_attributes = WP_Block_Supports::get_instance()->apply_block_supports( $current_parsed_block );
+
+	if ( empty( $new_attributes ) && empty( $extra_attributes ) ) {
+		return '';
+	}
+
+	// This is hardcoded on purpose.
+	// We only support a fixed list of attributes.
+	$attributes_to_merge = array( 'style', 'class' );
```

Comment:
I don't remember off the top of my head how params that are objects/arrays are documented, but this seems worth adding in the PHPDoc (although as minimal as `@param array $extra_attributes Optional. Extra classes or styles to render on the block wrapper. Use class or style as the array key, respectively.`).

Reply:
Hm, good catch. Is this filtering out any elements in `$extra_attributes` that aren't `style` or `class`? I think we should fix that. Even if keep that restriction for `$new_attributes` (which makes sense) I think we should lift it for `$extra_attributes`, otherwise it IMO goes against developer expectations.

(edit: I had added this comment in wrong place)

Reply:
The reason it's limited to a set of known attributes is because merging attributes is different from one to another. For example the `id` should overwrite while  the `style` should concat.

Reply:
mmm, I was thinking only about documenting expectations. Is there a need for this function to pass-through attributes other than style and class? My thinking is that it'd be ok if any other element attributes the block author needs is handled by its own mechanisms in the `render_callback` function (so outside this function). So, as long as it's documented, I believe this is fine as it is.